platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,720 @@
+instant would be for birth-round 200.
+
+In a sense, while both events are created at the same wall-clock time, Alice's event is actually older, because of what
+Alice understood about the state of the world at the time the event was created.
+
+It is not possible for any one node to get much farther ahead of all other nodes, since the only way to have a newer
+birth-round is to advance the hashgraph, and that requires 2/3 of the network by stake weight. Therefore, in this
+example, Alice is not just 100 rounds behind Bob, she must be 100 rounds behind at least 2/3 of the network by
+stake weight, or, Bob is lying. He may create events at round 200, but not actually have a hashgraph that has advanced
+to that round.
+
+Each event must be validated using the roster associated with its birth-round. If Alice is far behind Bob, and she
+receives an event for a birth-round she doesn't have the roster for, then she cannot validate the event. So the Event
+Intake module will drop events with a birth-round too far in the future.
+
+If the event has been dropped, it is possible to fetch missing events from neighbors using Gossip or, if the node has
+fallen so far behind it can no longer fetch events from neighbors in gossip, then the node will reconnect.
+
+#### Self Events
+
+Events are not only given to the Event Intake system through gossip. Self events (those events created by the node
+itself) are also fed to Event Intake. These events **may** bypass some steps in the pipeline. For example, self-events
+do not need validation. Likewise, when replaying events from the pre-consensus event buffer, those checks are not needed
+(since they have already been proved valid and are in topological order).
+
+#### Neighbor Discipline
+
+During the validation process, invalid events are rejected, and this information is passed to the Sheriff module so the
+offending node may be disciplined. Note that the node to be disciplined will be the node that sent this bad event, not
+the creator. This information (which node sent the event) must be captured by Gossip and passed to Event Intake as part
+of the event metadata.
+
+#### Assigning Generations
+
+Several algorithms, including the hashgraph consensus algorithm and the tipset algorithm, require events to be part of
+a "generation". We use what is known as "local generation". Each node, when it starts, assigns "1" as the generation
+of the first event it encounters per event creator. The absolute numbers for the generations are not important, but
+their relative values are critical. The next generation will be the max generation of all parents, plus 1. So if Alice
+builds an event with parents `a4` and `b7`, the new event will be `a8`. Different nodes may assign different generations
+to the same events. The generation will be part of the event metadata, **not** part of `EventCore` (that is, it is not
+in the part of the event that is gossiped).
+
+Event Intake is responsible for assigning generations to events.
```

Comment:
Even though some preconsensus events might not reach consensus, thus breaking the `generation = max(parentGen) +1` definition, is it accurate to say that "preconsensus generations" of consensus events would still always strictly increase when topologically sorted?

I'm just wondering if this would allow us to use a single definition of local generation.

Reply:
The two different types of generations have different mathematical definitions. Preconsensus generation does not need to be deterministic from node to node, and can be computed with the strategy described above. Consensus generation is something computed one round at a time as rounds reach consensus. Each round that reaches consensus will restart at generation 0 (or 1 if we decide to start with 1). I'd suggest reaching out to Leemon if the algorithm I describe above needs further explanation.

Reply:
Removing this section, since it isn't necessary to get into for the sake of this proposal.